schism/compiler.ss

```diff
@@ -490,15 +458,58 @@
   (define (parse-body body env)
     (unless (pair? body) (error 'parse-body "Empty body"))
     (parse-begin (parse-expr (car body) env) (cdr body) env))
-  (define (parse-function function env)
-    (let* ((name (definition-name function))
-           (args (function-formals function))
+  ;; help parse definitions
+  ;; def = (define (name args ...) body)
+  ;;     | (define name body?)
+  (define (definition-name def)
+    (and (pair? def)
+         (eq? (car def) 'define)
+         (pair? (cdr def))
+         (if (pair? (cadr def))
+             (caadr def)
+             (cadr def))))
+  (define (function-definition? def)
+    (and (pair? def)
+         (eq? (car def) 'define)
+         (pair? (cdr def))
+         (or (pair? (cadr def))
+             (and (pair? (cddr def))
+                  (pair? (caddr def))
+                  (eq? 'lambda (caaddr def))))))
+  (define (function-formals def)
+    (if (pair? (cadr def))
+        (cdadr def)
+        (let ((def* (caddr def)))
+          (unless (eq? (car def*) 'lambda)
+            (trace-and-error def 'function-formals "expected lambda at caaddr"))
+          (cadr def*))))
+  (define (function-body def)
+    (if (pair? (cadr def))
+        (cddr def)
+        (let ((def* (caddr def)))
+          (unless (eq? (car def*) 'lambda)
+            (trace-and-error def 'function-formals "expected lambda at caaddr"))
+          (cddr def*))))
+  (define (define-function def env)
+    (let* ((name (definition-name def))
+           (args (function-formals def))
            (args* (map rename-var args))
-           (body (parse-body (function-body function)
+           (body (parse-body (function-body def)
                              (add-lexicals args args* env))))
       `(,(cons name args*) ,body)))
-  (define (parse-functions definitions env)
-    (map (lambda (fn) (parse-function fn env)) definitions))
+  (define (define-value def env)
+    (let* ((name (definition-name def))
+           (env (memp (lambda (n.d) (eq? name (car n.d))) env))
```

Comment:
This ensures top level variable definitions don't reference unbound variables

Reply:
Makes sense. It wouldn't hurt to leave this as a comment in the code.